libsolidity/ast/AST.cpp

```diff
@@ -480,6 +480,63 @@ bool VariableDeclaration::canHaveAutoType() const
 	return isLocalVariable() && !isCallableParameter();
 }
 
+bool VariableDeclaration::isEventParameter() const
+{
+	return dynamic_cast<EventDefinition const*>(scope()) != nullptr;
+}
+
+bool VariableDeclaration::isReferenceType() const
+{
+	solAssert(typeName()->annotation().type, "Can only be called after reference resolution");
+	return dynamic_cast<ReferenceType const*>(typeName()->annotation().type.get());
+}
+
+set<VariableDeclaration::Location> VariableDeclaration::allowedStorageLocations() const
+{
+	using Location = VariableDeclaration::Location;
+
+	if (!isReferenceType())
+		return set<Location>{ Location::Default };
+
+	if (isEventParameter())
+		return set<Location>{ Location::Default };
+	else if (isConstant())
+		return set<Location>{Location::Default, Location::Memory};
+	else if (isExternalCallableParameter())
+	{
+		auto const& contract = dynamic_cast<ContractDefinition const&>(
+				*dynamic_cast<Declaration const&>(*scope()).scope()
+		);
+		if (contract.isLibrary())
+			return set<Location>{ Location::CallData, Location::Default, Location::Storage };
```

Comment:
I think in both these cases, `default` can be removed, but if this requires too many changes in the tests, we can also do it in a subsequent PR.